### PR TITLE
GCI-2294 Aggregation logic for Certified Copies

### PR DESCRIPTION
**Pull request by kinpang-CH**
* Add new CertifiedCopyConfirmationMapper
* Add respective models and above factory to ConfirmationMapperFactory
* Implement call to factory and mapper inside EmailService
* Refactor EmailService code
* Add new property to determine date format
* Add new property inside applications.properties to obtain string for emailConfig
* Implement logic inside CertifiedCopyConfirmationMapper to map view form link

Resolves [GCI-2294](https://companieshouse.atlassian.net/browse/GCI-2294)

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/itemhandler/itemsummary/CertifiedCopyConfirmationMapper.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                            ))
+                        .withCompanyNumber(item.getCompanyNumber())
+                        .withFee("£" + item.getTotalItemCost())
+                        .withViewFormLink(String.format(emailConfig.getCertifiedCopy().getFilingHistoryFormLink(),
+                                item.getCompanyNumber(), filingHistoryDocument.getFilingHistoryId()))
+                    .build();
+                }).forEach(certifiedCopyEmailData::add);
+    }
+
+    private String mapDateFiledFormat(FilingHistoryDocument filingHistoryDocument) {
```
Comment (kinpang-CH) [inline src/main/java/uk/gov/companieshouse/itemhandler/itemsummary/CertifiedCopyConfirmationMapper.java:70]: Use DateConstants to pass in the pattern

File: src/main/resources/application.properties [hunk 2]
```diff
@@ -29,8 +29,17 @@ response.handler.maximumRetryAttempts = 5
...
 email.certificate.standardSubjectLine = CHS certificate order
 email.certificate.expressSubjectLine = CHS Certificate Same day order
-email.senderEmail = noreply@companieshouse.gov.uk
\ No newline at end of file
+
+# Certified copy email properties
+email.certifiedCopy.recipient = ${CERTIFIED_COPY_ORDER_CONFIRMATION_RECIPIENT}
+email.certifiedCopy.standardSubjectLine = CHS certified copy order
+email.certifiedCopy.expressSubjectLine = CHS Certified Document Same day order
+email.certifiedCopy.dateFiledFormat = dd MMM yyyy
```
Comment (kinpang-CH) [inline src/main/resources/application.properties:44]: Can be removed and made a constant

File: src/main/resources/application.properties [hunk 3]
```diff
@@ -29,8 +29,17 @@ response.handler.maximumRetryAttempts = 5
...
+email.certifiedCopy.filingHistoryFormLink = ${CHS_URL}/company/%s/filing-history/%s/document?format=pdf&download=0
```
Comment (kinpang-CH) [inline src/main/resources/application.properties:45]: Pass {CHS_URL} to the mapper and map the property there.

**Review by cgregoryCH — APPROVED** [review]

**Review by sthompsonCH — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/itemhandler/service/EmailService.java [hunk 4]
```diff
@@ -103,56 +51,34 @@ public EmailService(
...
-        return new OrderConfirmationAndEmail(confirmation, email);
+    private EmailSend createEmailSendBasedOnKind(DeliverableItemGroup itemGroup) throws JsonProcessingException {
+        EmailSend emailSend = new EmailSend();
+        if (ITEM_KIND_CERTIFIED_COPY.equals(itemGroup.getKind())) {
+            EmailMetadata<CertifiedCopyEmailData> emailMetadata = confirmationMapperFactory.getCertifiedCopyMapper().map(itemGroup);
+            emailSend.setAppId(emailMetadata.getAppId());
+            emailSend.setMessageType(emailMetadata.getMessageType());
+            emailSend.setData(objectMapper.writeValueAsString(emailMetadata.getEmailData()));
+        } else {
+            EmailMetadata<CertificateEmailData> emailMetadata = confirmationMapperFactory.getCertificateMapper().map(itemGroup);
```
Comment (dlgroves-ch) [inline src/main/java/uk/gov/companieshouse/itemhandler/service/EmailService.java:77]: Duplicate code; you'd be better off moving logic back into `sendOrderConfirmation` and passing in the mapper as an argument:

```java
EmailSend emailModel;
if (ITEM_KIND_CERTIFIED_COPY.equals(itemGroup.getKind())) {
    emailModel = mapEmailSend(itemGroup, confirmationMapperFactory.getCertifiedCopyMapper());
else if (ITEM_KIND_CERTIFICATE.equals(itemGroup.getKind())) {
    emailModel = mapEmailSend(itemGroup, confirmationMapperFactory.getCertificateMapper());
} else {
    throw new RuntimeException(String.format("Unknown item kind: [%s]", itemGroup.getKind()));
}
...
```

```java
private EmailSend createEmailSendBasedOnKind(DeliverableItemGroup itemGroup) throws JsonProcessingException {
    EmailSend emailSend = new EmailSend();
    EmailMetadata<CertifiedCopyEmailData> emailMetadata = confirmationMapperFactory.getCertifiedCopyMapper().map(itemGroup);
    emailSend.setAppId(emailMetadata.getAppId());
    emailSend.setMessageType(emailMetadata.getMessageType());
    emailSend.setData(objectMapper.writeValueAsString(emailMetadata.getEmailData()));
    emailSend.setEmailAddress(TOKEN_EMAIL_ADDRESS);
    emailSend.setMessageId(UUID.randomUUID().toString());
    emailSend.setCreatedAt(LocalDateTime.now().toString());
    return emailSend;
}
```

File: src/main/java/uk/gov/companieshouse/itemhandler/service/EmailService.java [hunk 5]
```diff
@@ -103,56 +51,34 @@ public EmailService(
...
-                EmailMetadata<CertificateEmailData> emailMetadata = confirmationMapperFactory.getCertificateMapper().map(itemGroup);
-                EmailSend emailSend = new EmailSend();
-                emailSend.setAppId(emailMetadata.getAppId());
-                emailSend.setMessageType(emailMetadata.getMessageType());
-                emailSend.setData(objectMapper.writeValueAsString(emailMetadata.getEmailData()));
-                emailSend.setEmailAddress(TOKEN_EMAIL_ADDRESS);
-                emailSend.setMessageId(UUID.randomUUID().toString());
-                emailSend.setCreatedAt(LocalDateTime.now().toString());
+            EmailSend emailSend = createEmailSendBasedOnKind(itemGroup);
+            emailSend.setEmailAddress(TOKEN_EMAIL_ADDRESS);
```
Comment (dlgroves-ch) [inline src/main/java/uk/gov/companieshouse/itemhandler/service/EmailService.java:55]: Lines 55-57 break encapsulation; mapping logic has leaked out of your mapping method into client code.

File: src/main/java/uk/gov/companieshouse/itemhandler/itemsummary/ConfirmationMapperFactory.java [hunk 6]
```diff
@@ -6,12 +6,19 @@
...
         this.certificateConfirmationMapper = certificateConfirmationMapper;
+        this.certifiedCopyConfirmationMapper = certifiedCopyConfirmationMapper;
     }
 
     public OrderConfirmationMapper<CertificateEmailData> getCertificateMapper() {
         return certificateConfirmationMapper;
     }
+
+    public OrderConfirmationMapper<CertifiedCopyEmailData> getCertifiedCopyMapper() {
+        return  certifiedCopyConfirmationMapper;
```
Comment (dlgroves-ch) [inline src/main/java/uk/gov/companieshouse/itemhandler/itemsummary/ConfirmationMapperFactory.java:22]: Extra space.

**Review by dlgroves-ch — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/companieshouse/itemhandler/itemsummary/ConfirmationMapperFactory.java (hunk 6 above)
Comment (kinpang-CH) [reply src/main/java/uk/gov/companieshouse/itemhandler/itemsummary/ConfirmationMapperFactory.java]: Resolved in latest commit.

File: src/main/java/uk/gov/companieshouse/itemhandler/service/EmailService.java (hunk 5 above)
Comment (kinpang-CH) [reply src/main/java/uk/gov/companieshouse/itemhandler/service/EmailService.java]: Resolved in latest commit.

File: src/main/java/uk/gov/companieshouse/itemhandler/service/EmailService.java (hunk 4 above)
Comment (kinpang-CH) [reply src/main/java/uk/gov/companieshouse/itemhandler/service/EmailService.java]: Resolved in latest commit.

**Review by cgregoryCH — APPROVED** [review]

**Review by dlgroves-ch — APPROVED** [review]

**Pull request merged**